File: 2016/6/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('\n');

const frequencyMaps = [];
for (let i = 0; i < input[0].length; i++) {
  frequencyMaps.push(new Map());
}

for (const line of input) {
  for (let i = 0; i < line.length; i++) {
    if (!frequencyMaps[i].has(line[i])) {
      frequencyMaps[i].set(line[i], 0);
    }
    frequencyMaps[i].set(line[i], frequencyMaps[i].get(line[i]) + 1);
  }
}

let res = '';
let res2 = '';

for (const freqMap of frequencyMaps) {
  const freqArray = [];
  for (const [letter, n] of freqMap) {
    freqArray.push([letter, n]);
  }
  freqArray.sort((a, b) => b[1] - a[1]);
  res += freqArray[0][0];
  res2 += freqArray[freqArray.length - 1][0];
}

console.log(res);
console.log(res2);

File: 2016/6/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	fmt.Println(solve(puzzleInput))
}

func solve(puzzleInput []string) (string, string) {
	positionCountsMap := make(map[int]map[rune]int, len(puzzleInput[0]))
	for i := range len(puzzleInput[0]) {
		positionCountsMap[i] = make(map[rune]int, 26)
	}
	for _, s := range puzzleInput {
		for i, ch := range s {
			positionCountsMap[i][ch]++
		}
	}
	var sb1 strings.Builder
	var sb2 strings.Builder
	var maxCh, minCh rune
	var maxN, minN int
	for i := range len(puzzleInput[0]) {
		// unexpected order when using `for _, countMap := range positionCountsMap`
		// have to force it to iterate in numeric order
		countMap := positionCountsMap[i]
		maxN = 0
		minN = math.MaxInt
		for ch, n := range countMap {
			if n > maxN {
				maxN = n
				maxCh = ch
			}
			if n < minN {
				minN = n
				minCh = ch
			}
		}
		sb1.WriteRune(maxCh)
		sb2.WriteRune(minCh)
	}
	return sb1.String(), sb2.String()
}
